File: backend/app/models/meal_plan.py
"""Modele ORM planu posiłków i wpisów."""

from __future__ import annotations

import uuid
from datetime import date, datetime
from decimal import Decimal

from sqlalchemy import (
    Date,
    DateTime,
    ForeignKey,
    Integer,
    Numeric,
    String,
    func,
)
from sqlalchemy.dialects.postgresql import JSON, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.session import Base


class MealPlan(Base):
    """Plan posiłków na określony okres."""

    __tablename__ = "meal_plans"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=func.gen_random_uuid(),
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
    )
    store_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("stores.id", ondelete="CASCADE"),
        nullable=False,
    )
    start_date: Mapped[date] = mapped_column(Date, nullable=False)
    duration_days: Mapped[int] = mapped_column(Integer, nullable=False)
    meals_per_day: Mapped[int] = mapped_column(
        Integer, default=3, nullable=False
    )
    status: Mapped[str] = mapped_column(
        String(50),
        default="draft",
        nullable=False,
        comment="draft | active | completed | archived",
    )
    preferences: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

    # ── Relacje ──────────────────────────────────────────────────
    user: Mapped["User"] = relationship("User", back_populates="meal_plans")
    store: Mapped["Store"] = relationship("Store")
    entries: Mapped[list[MealPlanEntry]] = relationship(
        "MealPlanEntry", back_populates="meal_plan", cascade="all, delete-orphan"
    )
    shopping_list: Mapped["ShoppingList | None"] = relationship(
        "ShoppingList", back_populates="meal_plan", uselist=False
    )

    def __repr__(self) -> str:
        return f"<MealPlan {self.id} user={self.user_id} {self.start_date}>"


class MealPlanEntry(Base):
    """Wpis w planie posiłków — przypisanie przepisu do dnia i slotu."""

    __tablename__ = "meal_plan_entries"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=func.gen_random_uuid(),
    )
    meal_plan_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("meal_plans.id", ondelete="CASCADE"),
        nullable=False,
    )
    recipe_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("recipes.id", ondelete="CASCADE"),
        nullable=False,
    )
    day_number: Mapped[int] = mapped_column(Integer, nullable=False)
    meal_slot: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
        comment="śniadanie | obiad | kolacja | przekąska",
    )
    servings_multiplier: Mapped[Decimal] = mapped_column(
        Numeric(5, 2), default=Decimal("1.00"), nullable=False
    )

    # ── Relacje ──────────────────────────────────────────────────
    meal_plan: Mapped[MealPlan] = relationship(
        "MealPlan", back_populates="entries"
    )
    recipe: Mapped["Recipe"] = relationship("Recipe", back_populates="meal_plan_entries")

    def __repr__(self) -> str:
        return f"<MealPlanEntry day={self.day_number} slot={self.meal_slot!r}>"

File: frontend/lib/models/meal_plan.dart
import 'recipe.dart';

class MealPlanEntry {
  final String id;
  final String mealPlanId;
  final String recipeId;
  final Recipe recipe;
  final int dayNumber;
  final String mealSlot;
  final double servingsMultiplier;

  MealPlanEntry({
    required this.id,
    required this.mealPlanId,
    required this.recipeId,
    required this.recipe,
    required this.dayNumber,
    required this.mealSlot,
    required this.servingsMultiplier,
  });

  factory MealPlanEntry.fromJson(Map<String, dynamic> json) {
    return MealPlanEntry(
      id: json['id'] as String,
      mealPlanId: json['meal_plan_id'] as String? ?? '',
      recipeId: json['recipe_id'] as String? ?? '',
      recipe: Recipe.fromJson(json['recipe'] as Map<String, dynamic>),
      dayNumber: json['day_number'] as int? ?? 1,
      mealSlot: json['meal_slot'] as String? ?? json['meal_type'] as String? ?? 'obiad',
      servingsMultiplier: (json['servings_multiplier'] as num? ?? 1.0).toDouble(),
    );
  }
}

class MealPlan {
  final String id;
  final String userId;
  final String storeId;
  final String? startDate;
  final int durationDays;
  final int mealsPerDay;
  final String status; // 'draft', 'active', 'completed', 'archived'
  final double? estimatedMinBudget;
  final List<MealPlanEntry> entries;
  final DateTime createdAt;

  MealPlan({
    required this.id,
    required this.userId,
    required this.storeId,
    this.startDate,
    required this.durationDays,
    required this.mealsPerDay,
    required this.status,
    this.estimatedMinBudget,
    required this.entries,
    required this.createdAt,
  });

  factory MealPlan.fromJson(Map<String, dynamic> json) {
    return MealPlan(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      storeId: json['store_id'] as String,
      startDate: json['start_date'] as String?,
      durationDays: json['duration_days'] as int? ?? 3,
      mealsPerDay: json['meals_per_day'] as int? ?? 3,
      status: json['status'] as String? ?? 'draft',
      estimatedMinBudget: (json['estimated_min_budget'] as num?)?.toDouble(),
      entries: (json['entries'] as List<dynamic>?)
              ?.map((e) => MealPlanEntry.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  List<MealPlanEntry> entriesForDay(int day) {
    final list = entries.where((e) => e.dayNumber == day).toList();
    // Sortujemy posiłki według standardowego porządku dnia
    final order = {'śniadanie': 0, 'obiad': 1, 'kolacja': 2, 'przekąska': 3};
    list.sort((a, b) {
      final oa = order[a.mealSlot.toLowerCase()] ?? 99;
      final ob = order[b.mealSlot.toLowerCase()] ?? 99;
      return oa.compareTo(ob);
    });
    return list;
  }
}

class MealPlanGenerateRequest {
  final String storeId;
  final int durationDays;
  final int mealsPerDay;
  final double? maxBudget;
  final int? targetKcal;
  final int? householdSize;
  final Map<String, dynamic>? preferences;

  MealPlanGenerateRequest({
    required this.storeId,
    required this.durationDays,
    required this.mealsPerDay,
    this.maxBudget,
    this.targetKcal,
    this.householdSize,
    this.preferences,
  });

  Map<String, dynamic> toJson() => {
        'store_id': storeId,
        'duration_days': durationDays,
        'meals_per_day': mealsPerDay,
        'max_budget': maxBudget,
        'target_kcal': targetKcal,
        'household_size': householdSize,
        'preferences': preferences,
      };
}
